### PR TITLE
WIP: Create exercises skeleton, take 2: the electric bugaloo

### PR DESCRIPTION
**Pull request by tristan957**
- [x] Create config tests
- [x] Update documentation
- [x] Create db schema
- [ ] Create initial db models
- [ ] Create `GET /exercises/:id`
- [ ] Create `POST/GET /exercises`
- [ ] Create Neo4j tests with seeded data

### CODE REVIEW

**Comment by tristan957** [comment]
Disregard build failures. It is expecting python files, but I have not committed any due to their unfinished nature at the moment

File: README.md [hunk 1]
```diff
@@ -8,9 +8,16 @@ Contains the source code for the thesis completed under Dr. Philip B. Yasskin
...
+docker-compose -f docker-compose-dev.yml run exercises flake8 project # code quality with flake8 for project
 docker-compose -f docker-compose-dev.yml build # build
 docker-compose -f docker-compose-dev.yml up -d # bring up
 docker-compose -f docker-compose-dev.yml logs # view logs
-docker-compose -f docker-compose-dev.yml run users flake8 project # code quality with flake8
 ```
+
+## Documentation
+
+* [Neo4j Python Driver](https://neo4j.com/docs/api/python-driver/current/)
```
Comment (JosephMart) [inline README.md]: what?

File: docker-compose-dev.yml [hunk 2]
```diff
@@ -41,3 +41,34 @@ services:
...
+    depends_on:
+      - exercises-db
+    links:
+      - exercises-db
+
+  exercises-db:
+    container_name: exercises-db
+    image: neo4j:3.5
+    ports:
+      - 7474:7474 # http
```
Comment (JosephMart) [inline docker-compose-dev.yml]: map these different than source machine i.e `7475:7474`

File: README.md (hunk 1 above)
Comment (tristan957) [reply README.md]: Bookmarks to documentation!

Comment (JosephMart) [reply README.md]: I don't think that is readme worthy. Add it to the drive under resources I guess if you really want

Comment (tristan957) [reply README.md]: SAD! Removed in https://github.com/JosephMart/exercise-dependency-web/pull/2/commits/bc9a007e27fb10056d02bb1ec2c472120b27401a

Comment (JosephMart) [reply README.md]: https://github.com/JosephMart/exercise-dependency-web/pull/2/commits/bc9a007e27fb10056d02bb1ec2c472120b27401a

File: docker-compose-dev.yml [hunk 3]
```diff
@@ -62,3 +62,34 @@ services:
...
+      - DATABASE_URL=bolt://exerises-db:7687
+      - DATABASE_USERNAME=neo4j
+      - DATABASE_PASSWORD=neo4j
+    depends_on:
+      - exercises-db
+    links:
+      - exercises-db
+
+  exercises-db:
+    container_name: exercises-db
```
Comment (JosephMart) [inline docker-compose-dev.yml:92]: no

File: docker-compose-dev.yml [hunk 4]
```diff
@@ -62,3 +62,34 @@ services:
...
+    image: neo4j:3.5
```
Comment (JosephMart) [inline docker-compose-dev.yml]: no

**Comment by tristan957** [comment]
User model is done for the most part. Left are all other models.

**Comment by tristan957** [comment]
What is done is basically model properties. I am sure more will be added as time goes on, and we think about them more. They do not currently have any relationships whatsoever. This is where my knowledge is lacking in Neo4j. Tomorrow I am going to create a test user, and 2 textbooks they have written. Using that data I want to see how the JSON gets returned to me using different Cypher queries. I imagine it will hopefully be something similar to
```json
{
  "id": 1,
  "user_id": 1,
  ...,
  "textbooks": [
    {
      "id": 2,
      "title": "MYMA"
    }
  ]
}

File: services/exercises/project/db/domain/nodes/exercise.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        self.question = question
+        self.solution = solution
+        self.created_at = datetime.strptime(created_at, date_fmt_str)
+        self.updated_at = datetime.strptime(updated_at, date_fmt_str)
+
+    def serialize(self) -> Dict[str, Union[str, int, float, None]]:
+        """Serialize an `Exercise` object into a `dict`"""
+        return {
+            'id': self.id,
+            'question': self.question,
```
Comment (JosephMart) [inline services/exercises/project/db/domain/nodes/exercise.py:33]: There is also also answer and hint

File: services/exercises/project/db/domain/nodes/topic.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    `title`: title of the topic
+    """
+
+    title: str = None
+    created_at: datetime = None
+    updated_at: datetime = None
+
+    def __init__(self, id: Optional[int, None], title: str, created_at=datetime.now(), updated_at=datetime.now()):
+        Entity.__init__(self, id)
+        self.title = title
```
Comment (JosephMart) [inline services/exercises/project/db/domain/nodes/topic.py:20]: name > title

File: services/exercises/project/db/domain/nodes/textbook.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        Entity.__init__(self, id)
+        self.title = title
+        self.created_at = datetime.strptime(created_at, date_fmt_str)
+        self.updated_at = datetime.strptime(updated_at, date_fmt_str)
+
+    def serialize(self) -> Dict[str, Union[str, int, float, None]]:
+        """Serialize a `Textbook` object into a `dict`"""
+        return {
+            'id': self.id,
+            'title': self.title,
```
Comment (JosephMart) [inline services/exercises/project/db/domain/nodes/textbook.py:31]: name > title

File: services/exercises/project/db/domain/nodes/user.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 from project.db.domain.entity import Entity
 from project.db.domain import date_fmt_str
 from datetime import datetime
-from typing import Dict, Union
+from typing import Dict, Union, Optional
 
 
 class User(Entity):
```
Comment (JosephMart) [inline services/exercises/project/db/domain/nodes/user.py]: Is this supposed to duplicate users? I don't think this needs to exists in the graph db. I think the users_ids from the auth service should be stored on the textbook

Comment (tristan957) [reply services/exercises/project/db/domain/nodes/user.py]: hmm that is a better idea I think. I will change the model appropriately

File: services/exercises/project/db/domain/nodes/textbook.py (hunk 7 above)
Comment (tristan957) [reply services/exercises/project/db/domain/nodes/textbook.py:31]: I'll make the change, but I think title is better for textbooks.

File: services/exercises/project/db/domain/nodes/exercise.py (hunk 5 above)
Comment (tristan957) [reply services/exercises/project/db/domain/nodes/exercise.py:33]: is answer different from solution?

**Comment by tristan957** [comment]
Should title be changed to name for chapter as well?

**Comment by JosephMart** [comment]
yes

**Comment by JosephMart** [comment]
actually, i guess you can leave them all title

**Comment by JosephMart** [comment]
what does the payload look like?

**Comment by tristan957** [comment]
I have been working on `GET /textbooks?user_id=X`. I want the payload to look something like
```json
{
  "textbooks": [
    {
      "id": Y,
      "chapters": [
        {
          "id": X,
          "name": "MYMA",
          "children": []
        }
      ]
    }
  ]
}
```

**Comment by tristan957** [comment]
Should a textbook have the ability for mulitple authors?

**Comment by JosephMart** [comment]
Why do chapters have children key and not the came chapter key? I assume title will be in the textbooks object as well? Authors is not a thing, just user that have access to a book. Many users can have access to a textbook.

**Comment by JosephMart** [comment]
Can the textbooks payload just be textbooks with no chapter info?

Fetching that info should be a separate payload/endpoint

**Comment by tristan957** [comment]
Pretty sure your example JSON in slack had `children`. I will change it to chapters. I will also restrict this endpoint to only return the textbook props

**Comment by JosephMart** [comment]
I am saying they should be the same. both should be children or both should be chapters. They are the samething, why call it two different things

**Comment by JosephMart** [comment]
I prefer calling it children and then exercises having an isExercise flag

**Comment by tristan957** [comment]
For the record, I am not a fan of this children approach with the `isExercise` flag.

**Pull request closed without merging**